Add matched route instance as request attribute

othercorey opened the pull request:
Closes https://github.com/cakephp/cakephp/issues/16155

The route sets `_route` to itself when parsing and RouterMiddleware consumes it while setting the `route` attribute.

This will add `route` to the list of reserved request attributes.

othercorey committed:
Add matched route instance as request attribute

saeideng reviewed (commented):
Looks good

ADmad commented:
This comment can now be addressed https://github.com/cakephp/cakephp/pull/16151/files#r758032176.

ADmad commented:
To be clear the instance can now be directly used in Router::reverseToArray() instead of iterating the collection to find a match based on template string.

ADmad committed:
Optimize Router::reverseToArray().

It now uses the route instance from the request if available
instead of iterating the routes collection of find a match based
on the route template string.

ADmad commented:
Updated `Router::reverseToArray()`.

markstory reviewed (commented):
Looks good to me.
